src/reduceBy.js

```diff
@@ -0,0 +1,55 @@
+var _curryN = require('./internal/_curryN');
+var _reduce = require('./internal/_reduce');
+
+
+/**
+ * Groups the elements of the list according to the result of calling
+ * the String-returning function `keyFn` on each element and reduces the elements
+ * of each group to a single value via the reducer function `valueFn`.
+ *
+ * This function is basically a more general `groupBy` function.
+ *
+ * @func
+ * @memberOf R
+ * @since v0.20.0
+ * @category List
+ * @sig (b -> String) -> ((a, b) -> a) -> a -> [b] -> {String: a}
+ * @param {Function} keyFn The function that maps the list's element into a key.
+ * @param {Function} valueFn The function that reduces the elements of each group to a single
+ *        value. Receives two values, accumulator for a particular group and the current element.
+ * @param {*} acc The (initial) accumulator value for each group.
+ * @param {Array} list The array to group.
+ * @return {Object} An object with the output of `keyFn` for keys, mapped to the output of
+ *         `valueFn` for elements which produced that key when passed to `keyFn`.
+ * @see R.groupBy, R.reduce
+ * @example
+ *
+ *      var byGrade = R.reduceBy(function(student) {
+ *        var score = student.score;
+ *        return score < 65 ? 'F' :
+ *               score < 70 ? 'D' :
+ *               score < 80 ? 'C' :
+ *               score < 90 ? 'B' : 'A';
+ *      });
+ *      var namesByGrade = byGrade(function(acc, student) {
+ *        return acc.concat(student.name);
+ *      }, []);
+ *      var students = [{name: 'Lucy', score: 92},
+ *                      {name: 'Drew', score: 85},
+ *                      // ...
+ *                      {name: 'Bart', score: 62}];
+ *      namesByGrade(students);
+ *      // {
+ *      //   'A': ['Lucy'],
+ *      //   'B': ['Drew']
+ *      //   // ...,
+ *      //   'F': ['Bart']
+ *      // }
+ */
+module.exports = _curryN(4, [], function reduceBy(keyFn, valueFn, valueAcc, list) {
+  return _reduce(function(acc, elt) {
+    var key = keyFn(elt);
+    acc[key] = valueFn(acc[key] || valueAcc, elt);
```

Comment:
I believe this use of `||` to be unsound (as it so often is in JavaScript). If `key` is a value such as `'toString'` or `'valueOf'` this check will erroneously pass. Let's use our internal `_has` function to perform this check instead.